src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+        "      <arg name=\"queue_pos\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"queue_total\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"partition_name\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"partition_pos\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"partition_total\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"message\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"progress\" type=\"i\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <signal name=\"ProgressChanged\">\n"
+        "    </signal>\n"
+        "    <property name=\"running\" type=\"b\" access=\"read\">\n"
+        "    </property>\n"
+        "  </interface>\n"
+        "</node>\n"
+    "", &error);
+
+    if (!node_info) {
+        SFMF_FAIL("Could not compile D-Bus XML: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    ufs->object_registration = g_dbus_connection_register_object(ufs->system_bus, UFS_DBUS_PATH,
+            g_dbus_node_info_lookup_interface(node_info, UFS_DBUS_INTERFACE),
+            &upgrade_factory_snapshot_vtable, ufs, NULL, &error);
+
+    if (!ufs->object_registration) {
+        SFMF_FAIL("Could not register object on D-Bus: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    g_dbus_node_info_unref(node_info);
+}
+
+void upgrade_factory_snapshot_name_acquired_cb(GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    // Connect to D-Bus signals of the unpack utility
+    g_dbus_connection_signal_subscribe(ufs->system_bus, "org.sailfishos.sfmf.unpack",
+            "org.sailfishos.sfmf.unpack", NULL, "/", NULL, G_DBUS_SIGNAL_FLAGS_NONE,
+            upgrade_factory_snapshot_dbus_signal_cb, ufs, NULL);
+
+    // Schedule idle timer if no calls come in
+    upgrade_factory_snapshot_schedule_quit(ufs);
+}
+
+void upgrade_factory_snapshot_name_lost_cb(GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    if (!connection) {
+        SFMF_FAIL("Could not establish D-Bus connection\n");
+    } else {
+        SFMF_FAIL("D-Bus name lost: '%s'\n", name);
+    }
+}
+
+void upgrade_factory_snapshot_init(struct UpgradeFactorySnapshot *ufs)
+{
+    sfmf_cleanup_register(upgrade_factory_snapshot_cleanup_cb, ufs);
+
+    ufs->mainloop = g_main_loop_new(NULL, FALSE);
+    ufs->deploy_queue->callback_user_data = ufs;
+    ufs->cleanup_queue->callback_user_data = ufs;
+
+    for (ufs->deploy_queue->total=0; ufs->deploy_queue->tasks[ufs->deploy_queue->total].cmd; ufs->deploy_queue->total++);
+    for (ufs->cleanup_queue->total=0; ufs->cleanup_queue->tasks[ufs->cleanup_queue->total].cmd; ufs->cleanup_queue->total++);
+    for (ufs->partitions_length=0; ufs->partitions[ufs->partitions_length]; ufs->partitions_length++);
+
```

Comment:
I guess the first two are similar to the 3rd one, but can't be sure because github diffview cuts ends of lines... maybe break those long lines?